Implement Quantized Version of Threshold Function

paulshaoyuqiao opened the pull request:
Summary:
In this task, the quantized backend of the kernel is implemented for the threshold function, which clamps the entries in a tensor less than or equal to  a given threshold to be a specified value.

The corresponding Python implementation and unit test are also added.

Test Plan:
1. On a devserver, build PyTorch from source by running the command `buck build mode/dev //caffe2:torch`
2. Run the unit test throught the command
`buck test mode/dev //caffe2/test:quantization -- test_qthreshold`

Differential Revision: D21822446

z-a-f commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -631,6 +631,72 @@ void qclamp_kernel(
+          // dequantize
+          auto dx_vec = value_qx.dequantize(
+            input_scale_vec, input_zero_point_vec, input_scale_neg_zp_premul_vec);
+          for (int idx = 0; idx < dx_vec.size(); idx++) {
```
nit: Try using `++idx` instead of `idx++` whenever possible

paulshaoyuqiao replied:
I've added in the changes in the latest commit!

z-a-f commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -631,6 +631,72 @@ void qclamp_kernel(
+            auto cmp_to_threshold = dx_vec[idx] > threshold_vec;
+            if (cmp_to_threshold.zero_mask()) {
+              // blend
+              dx_vec[idx] = Vec::blendv(value_vec, dx_vec[idx], dx_vec[idx] > threshold_vec);
```
Isn't it supposed to be

```
dx_vec[idx] = Vec::blendv(value_vec, dx_vec[idx], cmp_to_threshold);
```

paulshaoyuqiao replied:
Yeah, I should have used `cmp_to_threshold` directly; added that change to the latest commit!

z-a-f commented on aten/src/ATen/native/quantized/cpu/qthreshold.cpp:
```diff
@@ -0,0 +1,45 @@
+  Tensor quantized_threshold_impl(
+      const Tensor& qx,
+      Scalar threshold,
+      Scalar value
+  ) {
```
nit: Be consistent

```
      Scalar value) {
```

paulshaoyuqiao replied:
I've added the change to the latest commit!

z-a-f commented on test/quantization/test_quantized_op.py:
```diff
@@ -418,6 +418,34 @@ def test_qtanh(self, X):
+                                           dtype=torch_type)
+
+        ops_under_test = {
+            'nn.quantized.functional': torch.nn.quantized.functional.threshold
```
I think we can also test `torch.nn.functonal.threshold` and `torch.threshold` in here, because they are supposed to support the quantized inputs

paulshaoyuqiao replied:
Sounds good! I've added both functions under the keys `nn.functional` and `native` to `ops_under_test`.

z-a-f commented on torch/nn/quantized/functional.py:
```diff
@@ -452,6 +452,25 @@ def hardswish(input, scale, zero_point):
+    if not input.is_quantized:
+        raise ValueError("Input to 'quantized.threshold' must be quantized!")
+    if (threshold is not None) != (value is not None):
+        raise ValueError("Either both or none of (threshold, value) must be specified!")
```
These arguments are non-optional, so both have to be specified

paulshaoyuqiao replied:
Sounds good! To be more specific, I've split the conditions instead into separate checks for `threshold` and `value` to make sure the inputs to both arguments must be specified.

z-a-f commented on torch/nn/quantized/functional.py:
```diff
@@ -452,6 +452,25 @@ def hardswish(input, scale, zero_point):
+    .. math::
+        \text{threshold}(x) = x \text{ if } x > \text{threshold else } \text{ value}
+
+    Args:
+        input: quantized input
+        threshold: the threshold value used to determine if :math:`x` should be
+        replaced by value
+        value: the value that is used in place of the original entries in x
```
Instead of putting the details about the arguments, we can just say `See :class:`~torch.nn.Threshold` for more details.`. That way in case the documentation changes, it will be the same everywhere

paulshaoyuqiao replied:
Got it! I've added in the change in the latest commit.

z-a-f commented on torch/nn/quantized/functional.py:
```diff
@@ -452,6 +452,25 @@ def hardswish(input, scale, zero_point):
+    r"""Applies the quantized version of the threshold function element-wise:
+
+    .. math::
+        \text{threshold}(x) = x \text{ if } x > \text{threshold else } \text{ value}
```
A better math equation would be

```latex
x = \begin{cases}
  x & \text{if~} x > \text{threshold} \\
  \text{value} & \text{otherwise}
\end{cases}
```

that will show 

![image](https://user-images.githubusercontent.com/4216323/83477429-13805700-a448-11ea-8f4d-a5dd53356a5d.png)

paulshaoyuqiao replied:
Sounds good!

z-a-f reviewed (changes requested):
Looks good -- minor changes

z-a-f reviewed (approved):
LGTM -- will leave it to other reviewers, in case they have comments

vkuzo commented:
(optional) since output quantization parameters are set equal to the input ones, would it make sense to do this in the quantized domain without doing dq -> float -> q?  Doesn't have to be in scope for this task.

Commit:
Implement Quantized Version of Threshold Function (#39352)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/39352

In this task, the quantized backend of the kernel is implemented for the threshold function, which clamps the entries in a tensor less than or equal to  a given threshold to be a specified value.

The corresponding Python implementation and unit test are also added.

Test Plan:
1. On a devserver, build PyTorch from source by running the command `buck build mode/dev //caffe2:torch`
2. Run the unit test throught the command
`buck test mode/dev //caffe2/test:quantization -- test_qthreshold`

Reviewed By: z-a-f

Differential Revision: D21822446

fbshipit-source-id: 4b3e86508642ec8c30d5087e3b76c026fac72c54
